plaidml/core/core.h

```diff
@@ -3,14 +3,51 @@
 #pragma once
 
 #include <cstring>
+#include <exception>
 #include <functional>
 #include <memory>
+#include <sstream>
 #include <stdexcept>
 #include <string>
 #include <vector>
 
 #include "plaidml/core/ffi.h"
 
+#if __has_include(<source_location>)
+#include <source_location>
+using edsl_source_location = std::source_location;
+#elif __cplusplus >= 201402 && __has_include(<experimental/source_location>)
+#include <experimental/source_location>
+using edsl_source_location = std::experimental::source_location;
+#else
+struct edsl_source_location {
+  static edsl_source_location current() { return edsl_source_location{}; }
+  constexpr uint_least32_t line() const noexcept { return 0; }
+  constexpr uint_least32_t column() const noexcept { return 0; }
+  constexpr const char* file_name() const noexcept { return "??"; }
+  constexpr const char* function_name() const noexcept { return "??"; }
+};
+#endif
+
+class ffi_exception : public std::exception {
+  std::string msg_;
+  edsl_source_location loc;
```

Comment:
Add `private:` just so that it's explicit (even though the default for `class` is to have private members).